### DISCUSSION TITLE
Installation of libraries from meson dependency

### DISCUSSION THREAD

**Discussion by Nic30** [mesonbuild/meson-python#419] [Q&A]
This may be similar to https://github.com/mesonbuild/meson-python/discussions/410 but my use-case is different.

I am using meson to build python extension (hdlConvertor) which depends on cmake builded library (antlr4 cpp runtime).
I can build and run the library, but during the installation the `pip3 install .` fails on this line https://github.com/mesonbuild/meson-python/blob/78def82ff1e6466afda91c69d7630cc89bc5177d/mesonpy/__init__.py#L163

```
      [251/251] Linking target hdlConvertor/_hdlConvertor.cpython-311-x86_64-linux-gnu.so
      + meson install --quiet --no-rebuild --destdir /home/nic30/Documents/workspace/hdlConvertor/.mesonpy-pjo43p7l/install
      
      meson-python: error: Could not map installation path to an equivalent wheel directory: '{prefix}/lib/libantlr4_static.a'
```
cmake builded antlr4 library contains
```cmake
  install(TARGETS antlr4_static
          EXPORT antlr4-targets
          ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR}
          LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
          RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR})
 ```
If I remove it, it works, but I would like to keep it and somehow tell cmake from meson that the target should not be installed or where it  shoud be installed.
`'CMAKE_SKIP_INSTALL_RULES': true, 'CMAKE_SKIP_INSTALL_ALL_DEPENDENCY': true,` does not seem to work.

During the build I can see
```
| CMake Warning in CMakeLists.txt:
| CMAKE_SKIP_INSTALL_RULES was enabled even though installation rules have
| been specified
```
But the install target still appears in intro-install_plan.json.
I guess that it is meson issue because wit set_install(false) it works https://github.com/mesonbuild/meson/blob/master/docs/markdown/CMake-module.md#cmake-options-object

Should I somehow specify the `INSTALL_PREFIX` for mentioned antrl4 library or I need to somehow exclude the library from installation?
I am trying to ask if the current meson-python supports the installation of such files
or I need to found out a way how to remove the file
or if I should implement this and send pull request to meson-python.

<strike>How can I get {prefix}, {includedir}, {libdir} and others in meson?</strike>
https://github.com/mesonbuild/meson/issues/10777 - mentioned constants must stay in `meson-info/intro-install_plan.json` in their native form.

However if I just add 
```
opt.add_cmake_defines({'CMAKE_INSTALL_PREFIX': '{libdir}')
```
I will end up with `"<project>/build/subprojects/antlr4-runtime/__CMake_build/{libdir}/lib/libantlr4_static.a",`
instead of just `{libdir}/lib/libantlr4_static.a`, any ideas?

**Comment by Mohataseem89** [comment]
It looks like this behavior is expected with the current `meson-python` design

`meson-python` only installs files that can be mapped to valid wheel locations (pure Python modules, extension modules headers, and required runtime shared libraries). static libraries like `libantlr4_static.a` do not have a well defined destination inside a wheel so the backend raises:
```
BuildError: Could not map installation path to an equivalent wheel directory
```


That’s why removing the `install()` rule makes the build succeed the file is no longer part of Meson’s install plan.

At the moment there is not a supported mechanism in `meson-python` to install arbitrary extra files or inject custom destinations. If a dependency provides a static archive, typical approaches are:

   -avoid installing it at all (e.g., `set set_install(false)` on the imported CMake target), because a static archive is normally linked into the     extension and does not need to ship in the wheel;
  
   -or build the dependency as a Meson subproject and configure installation yourself there.

`CMAKE_INSTALL_PREFIX` doesn’t help here because the substitution variables like `{prefix}`, `{libdir}` are placeholders internal to Meson’s install_plan file and are intentionally not expanded during wheel layout.

So today the recommended solution is:
  -disable installation of the CMake target (using `set_install(false)` in Meson’s CMake module),
  or
  -don’t export the archive as part of installation at all, since wheels don’t distribute `.a` archives.
  -There is no direct support in meson-python to map these files to wheel locations, so avoiding the install rule is the cleanest answer right now.

Hope that clarifies the current status.

**Comment by Mohataseem89** [comment]
It is expected behavior with the current design of meson-python.
The backend only knows how to install files that map cleanly into wheel locations (Python modules, extension modules, headers, and required shared libraries). A static archive like `libantlr4_static.a` has no valid destination inside a wheel, so it ends up in Meson’s `install_plan.json` with a `{prefix}/lib/...` location and meson-python raises:
```
BuildError: Could not map installation path to an equivalent wheel directory
```
Removing the `install()` rule makes the build succeed because the file is no longer in the install plan.


Right now there’s no meson-python mechanism to install arbitrary extra files or override the mapping. typical solutions in this situation are:
    - Do not install the static library at all it is normally linked into the extension and shouldn’t ship inside a wheel.
    - When importing via Meson’s CMake module, set `set_install(false)` on the target.
    - Or build the dependency as a Meson subproject and avoid exporting the `.a` archive for installation.

Changing `CMAKE_INSTALL_PREFIX` doesn’t help, because placeholders like `{prefix}`, `{libdir}`, etc. are Meson’s own substitution tokens and deliberately remain unexpanded in `intro-install_plan.json`.

So at the moment the clean fix is simply to avoid installing the static archive, since wheels aren’t meant to distribute `.a` files and meson-python does not support mapping them.

hope this help.
